tests/Runner/TestSuiteSorterTest.php

```diff
@@ -7,78 +7,322 @@
+                ],
+                ['testThree', 'testFive', 'testFour', 'testTwo', 'testOne']],
+
+            // Demonstrate a limit of the dependency resolver: after sorting defects to the front,
+            // the resolver will mark testFive done before testThree because of dependencies
+            'default resolve, testThree skipped, testFive fast' => [
+                TestSuiteSorter::ORDER_DEFAULT,
+                self::RESOLVE_DEPENDENCIES,
+                [
+                    'testOne'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testTwo'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testThree' => ['state' => BaseTestRunner::STATUS_SKIPPED, 'time' => 0],
+                    'testFour'  => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testFive'  => ['state' => BaseTestRunner::STATUS_SKIPPED, 'time' => 1],
+                ],
+                ['testFive', 'testOne', 'testTwo', 'testThree', 'testFour']],
+
+            // Torture test
+            // - incomplete TestResultCache
+            // - skipped testThree: will move it to the front as far as possible
+            // - testOne and testTwo are required before testThree, but can be reversed locally
+            // - testFive is independent will remain reversed up front
+            'reverse resolve, testThree skipped' => [
+                TestSuiteSorter::ORDER_REVERSED,
+                self::RESOLVE_DEPENDENCIES,
+                [
+                    'testOne'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testTwo'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testThree' => ['state' => BaseTestRunner::STATUS_SKIPPED, 'time' => 1],
+                ],
                 ['testFive', 'testTwo', 'testOne', 'testThree', 'testFour']],
+
+            // Make sure the dependency resolver is not confused by failing tests.
+            // Scenario: Four has a @depends on Three and fails. Result: Three is still run first
+            // testFive also fails but can be moved around freely and will be up front.
+            'depends first, then defects' => [
+                TestSuiteSorter::ORDER_DEFAULT,
+                self::RESOLVE_DEPENDENCIES,
+                [
+                    'testOne'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testTwo'   => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testThree' => ['state' => BaseTestRunner::STATUS_PASSED, 'time' => 1],
+                    'testFour'  => ['state' => BaseTestRunner::STATUS_FAILURE, 'time' => 1],
+                    'testFive'  => ['state' => BaseTestRunner::STATUS_FAILURE, 'time' => 1],
+                ],
+                ['testFive', 'testOne', 'testTwo', 'testThree', 'testFour']],
```

Comment:
:-1:
`testFour` and `testFive` failed, so I want to have them as soon as possible, with respect to their dependencies.

I would expect that order, instead:
`['testFive', 'testThree', 'testFour', 'testOne', 'testTwo']`
or
`['testThree', 'testFour', 'testFive', 'testOne', 'testTwo']`

Reply:
- `testThree` has dependencies on `testOne` and `testTwo`
- `testFour` has a dependency on `testThree`

what happens:
1. after sorting for defects the order is: `4` (failed), `5` (failed), `1`, `2`, `3`
2. running it like this would skip `4` because it needs to run `3` first
3. same story for `3`: it needs both `1` and `2`
4. the dependency resolver fixes this quietly for us: `5` can stay up front (=defects FIRST) but `4` goes behind `3` which goes behind `1,2` (=defects ASAP ;-)

Reply:
ha, ok, your first 2 points were not part of description (while dependencies of testThree for TestFour was)

then, all good :+1:

Reply:
Good but not good enough :) I will document the use of `MultiDependencyTest` in the dataproviders to avoid some of the confusion.